program/src/processor.rs

```diff
@@ -183,71 +191,69 @@ pub fn process_deposit(
     );
 
     lido.metrics.deposit_amount.observe(amount)?;
-    lido.save(accounts.lido)
+    lido.save(*accounts.lido.data.borrow_mut())
 }
 
 pub fn process_stake_deposit(
     program_id: &Pubkey,
     amount: Lamports,
+    validator_index: u32,
+    maintainer_index: u32,
     raw_accounts: &[AccountInfo],
 ) -> ProgramResult {
-    let accounts = StakeDepositAccountsInfo::try_from_slice(raw_accounts)?;
+    let accounts = StakeDepositAccountsInfoV2::try_from_slice(raw_accounts)?;
 
-    let mut lido = Lido::deserialize_lido(program_id, accounts.lido)?;
+    let lido = Lido::deserialize_lido(program_id, accounts.lido)?;
+    let lido_data = &mut *accounts.lido.data.borrow_mut();
 
-    lido.check_maintainer(accounts.maintainer)?;
+    lido.check_maintainer(lido_data, maintainer_index, accounts.maintainer)?;
     lido.check_reserve_account(program_id, accounts.lido.key, accounts.reserve)?;
     lido.check_stake_authority(program_id, accounts.lido.key, accounts.stake_authority)?;
     lido.check_can_stake_amount(accounts.reserve, amount)?;
 
-    let validator = lido.validators.get(accounts.validator_vote_account.key)?;
-
-    if !validator.entry.active {
-        msg!(
-            "Validator {} is inactive, new deposits are not allowed",
-            validator.pubkey
-        );
-        return Err(LidoError::StakeToInactiveValidator.into());
-    }
+    let mut validators = Lido::deserialize_account_list::<Validator>(lido_data)?;
```

Comment:
I think I mentioned it before, but I think it would be nice to package this together with `Lido::deserialize_lido`, so it would return a tuple of the Lido instance, and also the validator list and maintainer list. Then we know for sure that they are consistent with one another.

Reply:
O, now I get what you meant. The problem is that you can't mutably access `accounts.lido.data` simultaneously by maintainers and validators, the almighty borrow checker will not be happy. So we can't return them both from `deserialize_account_list`. So the only option is to access `accounts.lido.data` sequentially.

Reply:
> The problem is that you can't mutably access accounts.lido.data simultaneously by maintainers and validators, the almighty borrow checker will not be happy

You can, with [`slice::split_at_mut`](https://doc.rust-lang.org/std/primitive.slice.html#method.split_at_mut).

Reply:
I have done it multiple times with all tests passing in different variants, but my aesthetic sense tells me it looks ugly. Instead one line I should add 3 new:

```rust
-    let mut lido = Lido::deserialize_lido(program_id, accounts.lido)?;

+    let lido_data = &mut *accounts.lido.data.borrow_mut();
+    let mut slices = LidoSlices::new(lido_data)?;
+    let (mut lido, _, _) = Lido::deserialize_lido(program_id, accounts.lido, &mut slices)?;
```

where
```rust
pub struct LidoSlices<'a> {
    pub lido: &'a mut [u8],
    maintainers: &'a mut [u8],
    validators: &'a mut [u8],
}
```
So I quit it. Don't want to spend more of my time on syntactic sugar, although  they could be inconsistent as you mention here. But I don't think it's a problem at all.

Reply:
No, reverting it back. It is ugly.